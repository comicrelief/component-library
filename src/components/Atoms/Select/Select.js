import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Label from '../Label/Label';
import ErrorText from '../ErrorText/ErrorText';
import dropDownIcon from './assets/drop-down-dark-purple.svg';
import spacing from '../../../theme/shared/spacing';

const StyledSelect = styled.select`
  width: 100%;
  font-size: ${({ theme }) => theme.fontSize('m')};
  display: block;
  box-sizing: border-box;
  padding: 0 ${spacing('m')};
  padding-right: ${spacing('xl')};
  margin: 0;
  position: relative;
  height: 48px;
  font-weight: 400;
  font-family: ${({ theme }) => theme.fontFamilies(theme.font.regular)};
  background: ${({ theme }) => theme.color('grey_light')} url(${dropDownIcon})
    calc(100% - 1.5rem) 14px/20px 1.5rem no-repeat;
  border: 1px solid;
  border-color: ${({ theme, error }) => (error ? theme.color('red') : theme.color('grey_medium'))};
  box-shadow: none;
  appearance: none;
  color: ${({ theme, greyDescription, hasValue }) => (greyDescription && !hasValue ? 'grey' : theme.color('black'))};
  border-radius: 0.5rem;
  cursor: pointer;
  :focus {
    border: 1px solid ${({ theme }) => theme.color('grey_for_forms')};
  }

  @media ${({ theme }) => theme.allBreakpoints('M')} {
    max-width: 290px;
  }
`;

const Select = React.forwardRef(
  (
    {
      errorMsg,
      description,
      label,
      options,
      hideLabel = false,
      defaultValue = '',
      onChange = null,
      greyDescription = false,
      className = '',
      optional = false,
      ...rest
    },
    ref
  ) => {
    const [value, setValue] = useState('');

    return (
      <Label
        label={label}
        hideLabel={hideLabel}
        errorMsg={errorMsg}
        className={className}
        optional={optional}
      >
        <StyledSelect
          onChange={e => {
            setValue(e.currentTarget.value);
            if (onChange) {
              onChange(e);
            }
          }}
          {...rest}
          error={errorMsg}
          defaultValue={defaultValue}
          required={optional === false}
          hasValue={!!value}
          greyDescription={greyDescription}
          ref={ref}
        >
          <option disabled value="">
            {description}
          </option>
          {options.map(option => (
            <option key={option.value} value={option.value}>
              {option.displayValue}
            </option>
          ))}
        </StyledSelect>
        {errorMsg && <ErrorText size="error" weight="bold" data-test="error-message">{errorMsg}</ErrorText>}
      </Label>
    );
  }
);

Select.propTypes = {
  label: PropTypes.string.isRequired,
  hideLabel: PropTypes.bool,
  errorMsg: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
      displayValue: PropTypes.string.isRequired
    })
  ).isRequired,
  description: PropTypes.string.isRequired,
  defaultValue: PropTypes.string,
  onChange: PropTypes.func,
  greyDescription: PropTypes.bool,
  // className is needed so that styled(`Select`) will work
  // (as `rest` is not spread on the outermost component)
  className: PropTypes.string,
  optional: PropTypes.bool
};

export default Select;
