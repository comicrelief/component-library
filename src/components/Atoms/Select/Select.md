# Select

```js
<h4>DROPDOWN MENU</h4>
<Select
  label="Label"
  description="Please choose an option"
  errorMsg=""
  options={[
    { value: 'Option one', displayValue: 'The first option' },
    {
      value: 'Option two',
      displayValue: 'The second option'
    },
    { value: 'Option three', displayValue: 'The third option' },
    { value: 'Option four', displayValue: 'The fourth option' }
  ]}
/>

<h4>DROPDOWN MENU</h4>
<Select
  label="Label"
  errorMsg="This is an error message"
  description="Please choose an option"
  greyDescription
  options={[
    { value: 'Option one', displayValue: 'The first option' },
    {
      value: 'Option two',
      displayValue: 'The second option'
    },
    { value: 'Option three', displayValue: 'The third option' },
    { value: 'Option four', displayValue: 'The fourth option' }
  ]}
/>
```
