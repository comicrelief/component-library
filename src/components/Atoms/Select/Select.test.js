import React from 'react';
import 'jest-styled-components';
import renderWithTheme from '../../../hoc/shallowWithTheme';
import Select from './Select';

it('renders correctly', () => {
  const tree = renderWithTheme(
    <Select
      label="Label"
      description="Please choose an option"
      errorMsg=""
      options={[
        { value: 'Option one', displayValue: 'The first option' },
        {
          value: 'Option two',
          displayValue: 'The second option'
        },
        { value: 'Option three', displayValue: 'The third option' },
        { value: 'Option four', displayValue: 'The fourth option' }
      ]}
    />
  ).toJSON();

  expect(tree).toMatchInlineSnapshot(`
    .c1 {
      font-size: 1rem;
      line-height: 1rem;
      text-transform: inherit;
      font-weight: bold;
      line-height: normal;
      font-family: 'Montserrat',Helvetica,Arial,sans-serif;
    }

    .c2 {
      width: 100%;
      font-size: 1.25rem;
      display: block;
      box-sizing: border-box;
      padding: 0.8rem 1.5rem;
      padding-right: 4rem;
      margin: 0;
      position: relative;
      height: 48px;
      font-weight: 400;
      background: #F4F3F5 url(mock.asset) calc(100% - 1.5rem) 14px/20px 1.5rem no-repeat;
      border: 1px solid;
      border-color: #E1E2E3;
      box-shadow: none;
      -webkit-appearance: none;
      -moz-appearance: none;
      appearance: none;
      color: #000000;
      border-radius: 0.5rem;
      margin-top: 0.5rem;
      cursor: pointer;
    }

    .c2:focus {
      border: 1px solid #666;
    }

    .c0 {
      display: -webkit-box;
      display: -webkit-flex;
      display: -ms-flexbox;
      display: flex;
      -webkit-flex-direction: column;
      -ms-flex-direction: column;
      flex-direction: column;
      font-weight: bold;
      color: #5C5C5E;
    }

    @media (min-width:740px) {
      .c2 {
        max-width: 290px;
      }
    }

    <label
      className="c0"
    >
      <span
        className="c1 "
        color="inherit"
        size="s"
      >
        Label
      </span>
      <select
        className="c2"
        defaultValue=""
        onChange={[Function]}
      >
        <option
          disabled={true}
          value=""
        >
          Please choose an option
        </option>
        <option
          value="Option one"
        >
          The first option
        </option>
        <option
          value="Option two"
        >
          The second option
        </option>
        <option
          value="Option three"
        >
          The third option
        </option>
        <option
          value="Option four"
        >
          The fourth option
        </option>
      </select>
      
    </label>
  `);
});
