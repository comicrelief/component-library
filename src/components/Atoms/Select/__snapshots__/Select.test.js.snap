// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c1 {
  font-size: 1rem;
  line-height: 1rem;
  text-transform: inherit;
  font-weight: bold;
  line-height: normal;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c2 {
  width: 100%;
  font-size: 1.25rem;
  display: block;
  box-sizing: border-box;
  padding: 0.8rem 1.5rem;
  padding-right: 4rem;
  margin: 0;
  position: relative;
  height: 48px;
  font-weight: 400;
  background: #F4F3F5 url(mock.asset) calc(100% - 1.5rem) 14px/20px 1.5rem no-repeat;
  border: 1px solid;
  border-color: #E1E2E3;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #000000;
  border-radius: 0.5rem;
  margin-top: 0.5rem;
  cursor: pointer;
}

.c2:focus {
  border: 1px solid #666;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-weight: bold;
  color: #5C5C5E;
}

@media (min-width:740px) {
  .c2 {
    max-width: 290px;
  }
}

<label
  className="c0"
>
  <span
    className="c1 Select__LabelText-sc-1atp8e3-2"
    color="inherit"
    size="s"
  >
    Label
  </span>
  <select
    className="c2"
    defaultValue=""
    onChange={[Function]}
  >
    <option
      disabled={true}
      value=""
    >
      Please choose an option
    </option>
    <option
      value="Option one"
    >
      The first option
    </option>
    <option
      value="Option two"
    >
      The second option
    </option>
    <option
      value="Option three"
    >
      The third option
    </option>
    <option
      value="Option four"
    >
      The fourth option
    </option>
  </select>
  
</label>
`;
